engine/runtime/src/main/java/org/enso/interpreter/runtime/data/text/Text.java

```diff
@@ -213,63 +203,39 @@ String toDisplayString(
     return Core_Text_Utils.prettyPrint(str);
   }
 
-  private void setContents(String contents) {
-    assert length == -1 || length == contents.length();
-    this.contents = contents;
-  }
-
-  private void setFcdNormalized(boolean flag) {
-    if (flag) {
-      fcdNormalized = FcdNormalized.YES;
-    } else {
-      fcdNormalized = FcdNormalized.NO;
-    }
-  }
-
   @Override
   public String toString() {
-    Object c = this.contents;
-    if (c instanceof String s) {
-      return s;
-    } else {
-      return flattenIfNecessary(this);
-    }
+    return switch (this.contents) {
+        case String s -> s;
+        case ConcatRope r -> flattenAndSetContent(r);
+        case null, default -> throw new NullPointerException();
+    };
   }
 
   /**
    * Converts text to a Java String. For use outside of Truffle Nodes.
    *
-   * @param text the text to convert.
+   * @param c the content to flatten
    * @return the result of conversion.
    */
   @CompilerDirectives.TruffleBoundary
-  private static String flattenIfNecessary(Text text) {
-    LOCK.lock();
-    String result;
-    try {
-      Object c = text.contents;
-      if (c instanceof String s) {
-        result = s;
-      } else {
-        Deque<Object> workStack = new ArrayDeque<>();
-        StringBuilder bldr = new StringBuilder();
-        workStack.push(c);
-        while (!workStack.isEmpty()) {
-          Object item = workStack.pop();
-          if (item instanceof String) {
-            bldr.append((String) item);
-          } else {
-            ConcatRope rope = (ConcatRope) item;
-            workStack.push(rope.getRight());
-            workStack.push(rope.getLeft());
+  private String flattenAndSetContent(Object c) {
+    var workStack = new ArrayDeque<Object>();
+    StringBuilder bldr = new StringBuilder();
+    workStack.push(c);
+    while (!workStack.isEmpty()) {
+      switch (workStack.pop()) {
+          case String s -> bldr.append(s);
+          case ConcatRope rope -> {
+            workStack.push(rope.right());
+            workStack.push(rope.left());
           }
-        }
-        result = bldr.toString();
-        text.setContents(result);
+          case null, default -> throw new NullPointerException();
```

Comment:
IMO throwing NPE on `default` branch is misleading. We haven't encountered any null pointer in this case, but an unexpected class. So perhaps ClassCastException or just IllegalStateException would fit better?

I know it may seem like nitpicking, but when debugging, misleading exceptions do not help. I know we can follow the stack trace to find this place, but tbh I'd still be confused by the exception kind here.

Reply:
> misleading exceptions do not help

Hopefully the exception will be accompanied with a stack trace. Then it should point directly to the problem and not be misleading at all.